### PR TITLE
Use aligned memory if needed.

### PR DESCRIPTION
**Pull request by markdroth**
Fix `New<>()` and `Delete<>()` to use aligned memory allocation if needed.  This is an attempt to fix a ubsan problem reported by oss-fuzz.

I am not sure that the value of `ALIGNMENT_FOR_DEFAULT_ALLOCATION` is right for all platforms, but it seems to be right for glibc malloc.  I'm not sure if this needs a portability macro to be set differently on other platforms or not.

I am also not sure if this has performance implications.  It seems like it will cause us to allocate more memory, though, so it might.  A potential improvement would be to replace `gpr_malloc_aligned()` with `posix_memalign()`, but that too probably has portability implications.  For now, let's see what the microbenchmarks report.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                FILE SIZE
 ++++++++++++++ GROWING                                                  ++++++++++++++
  +0.4%     +16 src/core/ext/filters/client_channel/resolver_registry.cc     +16  +0.4%
       +26%     +11 [Unmapped]                                                   +11   +26%
      +5.2%      +5 grpc_core::ResolverRegistry::Builder::InitRegistry            +5  +5.2%

 -+-+-+-+-+-+-+ MIXED                                                    +-+-+-+-+-+-+-
  -0.0%     -16 [None]                                                       +16  +0.0%

  [ = ]       0 TOTAL                                                        +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by kcc** [comment]
Note: this has nothing to do with ubsan. 
If the compiler assumes 16-byte alignment it may use instructions that will SEGV on unaligned data.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/gprpp/memory.h [hunk 1]
```diff
@@ -27,18 +27,27 @@
 
 namespace grpc_core {
 
+// The alignment of memory returned by gpr_malloc().
+#define ALIGNMENT_FOR_DEFAULT_ALLOCATION 8
```
Comment (dgquintas) [inline src/core/lib/gprpp/memory.h]: we could use a `constexpr` for this. We should also specify the units as part of the name.

Comment (markdroth) [reply src/core/lib/gprpp/memory.h]: Done.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                FILE SIZE
 ++++++++++++++ GROWING                                                  ++++++++++++++
  +0.4%     +16 src/core/ext/filters/client_channel/resolver_registry.cc     +16  +0.4%
       +26%     +11 [Unmapped]                                                   +11   +26%
      +5.2%      +5 grpc_core::ResolverRegistry::Builder::InitRegistry            +5  +5.2%

 -+-+-+-+-+-+-+ MIXED                                                    +-+-+-+-+-+-+-
  +0.0%     +16 [None]                                                       -56  -0.0%

  +0.0%     +32 TOTAL                                                        -40  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                              cpu_time    real_time
-------------------------------------------------------------------------------------  ----------  -----------
BM_PumpStreamClientToServer<InProcess>/262144                                          +6%         +6%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/262144/2/1  -5%         -5%
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                FILE SIZE
 ++++++++++++++ GROWING                                                  ++++++++++++++
  +0.4%     +16 src/core/ext/filters/client_channel/resolver_registry.cc     +16  +0.4%
       +26%     +11 [Unmapped]                                                   +11   +26%
      +5.2%      +5 grpc_core::ResolverRegistry::Builder::InitRegistry            +5  +5.2%

 -+-+-+-+-+-+-+ MIXED                                                    +-+-+-+-+-+-+-
  +0.0%     +16 [None]                                                       -56  -0.0%

  +0.0%     +32 TOTAL                                                        -40  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Pull request merged**